pkg/middleware/auth/middleware.go

```diff
@@ -1,152 +1,217 @@
 	}
 
-	middlewareLogger.Debug(" Creating new auth middleware")
+	transportCfg := httptransport.TransportConfig{
+		Wallet:                 cfg.Wallet,
+		SessionManager:         sessionManager,
+		Logger:                 logger,
+		CertificatesToRequest:  cfg.CertificatesToRequest,
+		OnCertificatesReceived: cfg.OnCertificatesReceived,
+	}
 
-	t := httptransport.New(opts.Wallet, opts.SessionManager, opts.AllowUnauthenticated, opts.Logger, opts.CertificatesToRequest, opts.OnCertificatesReceived)
+	t := httptransport.New(transportCfg)
 
-	middlewareLogger.Debug(" transport created")
+	peerCfg := &auth.PeerOptions{
+		Wallet:                cfg.Wallet,
+		Transport:             t,
+		SessionManager:        sessionManager,
+		CertificatesToRequest: cfg.CertificatesToRequest,
+	}
+	peer := auth.NewPeer(peerCfg)
+	peer.ListenForCertificatesReceived(cfg.OnCertificatesReceived)
 
 	return &Middleware{
-		wallet:               opts.Wallet,
-		sessionManager:       opts.SessionManager,
+		peer:                 peer,
+		wallet:               cfg.Wallet,
+		sessionManager:       sessionManager,
 		transport:            t,
-		allowUnauthenticated: opts.AllowUnauthenticated,
+		allowUnauthenticated: cfg.AllowUnauthenticated,
 		logger:               middlewareLogger,
 	}, nil
 }
 
-// Handler returns standard http middleware
+// Handler returns a middleware handler function for BRC authentication
 func (m *Middleware) Handler(next http.Handler) http.Handler {
-	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
-		recorder := newResponseRecorder(w)
-		if req.Method == http.MethodPost && req.URL.Path == "/.well-known/auth" {
-			err := m.transport.HandleNonGeneralRequest(req, recorder)
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		wrappedWriter := httptransport.WrapResponseWriter(w)
+
+		ctx := context.WithValue(r.Context(), httptransport.RequestKey, r)
+		ctx = context.WithValue(ctx, httptransport.ResponseKey, wrappedWriter)
+		ctx = context.WithValue(ctx, httptransport.NextKey, func() {
+			next.ServeHTTP(wrappedWriter, r)
+		})
+
+		m.logger.Debug("Processing request",
+			slog.String("path", r.URL.Path),
+			slog.String("method", r.Method))
+
+		authMsg, err := httptransport.ParseAuthMessageFromRequest(r)
+		if err != nil {
+			m.logger.Error("Failed to parse auth message", slog.String("error", err.Error()))
+			http.Error(wrappedWriter, err.Error(), http.StatusBadRequest)
+			return
+		}
+
+		if authMsg == nil {
+			if m.allowUnauthenticated {
+				r = r.WithContext(context.WithValue(r.Context(), httptransport.IdentityKey, "unknown"))
+				next.ServeHTTP(wrappedWriter, r)
+				return
+			} else {
+				// BRC-104 requires 401 for authentication failures
+				http.Error(wrappedWriter, "Authentication required", http.StatusUnauthorized)
+				return
+			}
+		}
+
+		if authMsg.IdentityKey == nil && r.Header.Get("x-bsv-auth-identity-key") != "" {
```

Comment:
Maybe I'm missing something, but I believe that you're getting authMsg.IdentityKey from the same header in ParseAuthMessageFromRequest.

If there is a possiblity that it's not set in one of the if branches, then maybe the function should have this additional loading from header, instead of moving this logic here?

Reply:
Moved Identity key extraction logic into `ParseAuthMessageFromRequest` to handle all cases consistently.